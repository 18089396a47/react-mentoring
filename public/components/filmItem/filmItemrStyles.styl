.film-item
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  flex-basis: 0;
  margin: 0 40px 40px 0;
  height: 390px;
  list-style: none;
  color: #3e3e3e;

  &__poster
    width: 210px;

  &__info-wrapper
    padding: 10px 5px 0;

  &__title-wrapper
    display: flex;
    justify-content: space-between;
    align-items: flex-start;
    margin-bottom: 5px;
    line-height: 22px;

  &__title-link
    text-decoration: none;
    color: inherit;

  &__title
    font-weight: bold;
    font-size: 15px;
    text-transform: uppercase;

  &__year
    font-size: 12px;
    line-height: 12px;
    padding: 1px 3px;
    border: 1px solid #3e3e3e;
    flex-shrink: 0;

  &__category
    font-size: 12px;
