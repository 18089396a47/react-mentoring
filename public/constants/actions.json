{
  "searchStart": "SEARCH_START",
  "searchEnd": "SEARCH_END",
  "changeSearchInput": "CHANGE_SEARCH_INPUT",
  "updateSearchResults": "UPDATE_SEARCH_RESULTS"
}
