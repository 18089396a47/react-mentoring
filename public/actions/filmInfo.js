import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const searchFilmEnd = () => ({
    type: types.searchFilmEnd
});

export const searchFilmStart = (id, searchType) => ({
    type: types.searchFilmStart,
    id,
    searchType
});

function* fetchFilm({ id, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.movies.getById : theMovieDb.tv.getById;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            id
        }));

        yield put(searchFilmEnd());
        yield put(updateFilmInfo(JSON.parse(response)))
    } catch (e) {
        yield put(searchFilmEnd());
    }
}

function* watchFetchFilm() {
    yield takeLatest(types.searchFilmStart, fetchFilm);
}

export function* filmInfoSaga() {
    yield all([
        watchFetchFilm()
    ]);
}

export const updateFilmInfo = (data) => ({
    type: types.updateFilmInfo,
    data
});
