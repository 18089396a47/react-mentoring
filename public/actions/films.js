import * as types from '../constants/actions';

export const updateSearchResults = (data) => ({
    type: types.updateSearchResults,
    data: JSON.parse(data)
});
