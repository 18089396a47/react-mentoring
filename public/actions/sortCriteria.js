import * as types from '../constants/actions';

export const changeSortCriteria = (sortCriteria) => ({
    type: types.changeSortCriteria,
    sortCriteria
});
