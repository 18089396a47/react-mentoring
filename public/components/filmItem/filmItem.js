import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import './filmItemrStyles';
import SearchMovieField from '../searchMovieField/searchMovieField';
import InfoPanel from '../infoPanel/infoPanel';
import getYear from '../../helpers/getYear';
import { searchFilmStart, searchSimilarFilmStart } from '../../actions/';

const FilmItem = ({
    dispatch,
    searchType,
    show_title,
    release_year,
    category,
    id,
    poster
}) => {
    const searchFilm = () => {
        dispatch(searchFilmStart(id, searchType));
        dispatch(searchSimilarFilmStart(id));
    };

    return (
        <li className="film-item">
            <Link to={`/film/${id}`} onClick={searchFilm}>
                <img className="film-item__poster" src={poster} alt={show_title} />
            </Link>
            <div className="film-item__info-wrapper">
                <div className="film-item__title-wrapper">
                    <Link className="film-item__title-link" to={`/film/${show_title}`} onClick={searchFilm}>
                        <h3 className="film-item__title">{show_title}</h3>
                    </Link>
                    <span className="film-item__year">{getYear(release_year)}</span>
                </div>
                <span className="film-item__category">{category}</span>
            </div>
        </li>
    );
};

FilmItem.propTypes = {
    dispatch: PropTypes.func.isRequired,
    searchType: PropTypes.string.isRequired,
    show_title: PropTypes.string.isRequired,
    release_year: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    poster: PropTypes.string.isRequired
};

const mapStateToProps = (state) => {
    return {
        searchType: state.search.searchType
    };
};

export default connect(mapStateToProps)(FilmItem);
