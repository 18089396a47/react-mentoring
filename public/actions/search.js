import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import { updateSearchResults } from './films';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const changeSearchInput = (inputValue) => ({
    type: types.changeSearchInput,
    inputValue
});

export const searchQueryEnd = () => ({
    type: types.searchQueryEnd
});

export const searchQueryStart = (inputValue, searchType) => ({
    type: types.searchQueryStart,
    inputValue,
    searchType
});

function* fetchQuery({ inputValue, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.search.getMovie : theMovieDb.search.getTv;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            query: inputValue
        }));

        yield put(searchQueryEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchQueryEnd());
    }
};

function* watchFetchQuery() {
    yield takeLatest(types.searchQueryStart, fetchQuery);
}

export const searchSimilarFilmEnd = () => ({
    type: types.searchSimilarFilmEnd
});

export const searchSimilarFilmStart = (id) => ({
    type: types.searchSimilarFilmStart,
    id
});

function* fetchSimilarFilms({ id }) {
    try {
        const response = yield call(() => promiseWrapper(theMovieDb.movies.getSimilarMovies, {
            id
        }));

        yield put(searchSimilarFilmEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchSimilarFilmEnd());
    }
};

function* watchFetchSimilarFilms() {
    yield takeLatest(types.searchSimilarFilmStart, fetchSimilarFilms);
}

export function* searchSaga() {
    yield all([
        watchFetchSimilarFilms(),
        watchFetchQuery()
    ]);
}

export const changeSearchType = (searchType) => ({
    type: types.changeSearchType,
    searchType
});
