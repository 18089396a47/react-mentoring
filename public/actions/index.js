import { all } from 'redux-saga/effects';

import { filmInfoSaga } from './filmInfo';
import { searchSaga } from './search';

export default function* rootSaga() {
    yield all([
        filmInfoSaga(),
        searchSaga()
    ]);
}

export * from './search';
export * from './sortCriteria';
export * from './filmInfo';
export * from './films';
