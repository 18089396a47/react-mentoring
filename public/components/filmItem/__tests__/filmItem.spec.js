import React from 'react';
import renderer from 'react-test-renderer';
import { FilmItem } from '../filmItem';
import { searchFilmStart, searchSimilarFilmStart } from '../../../actions/';

const mockDigest = jest.fn();

jest.mock('react-router-dom', () => ({ Link: 'Link' }));
jest.mock('../../searchMovieField/searchMovieField', () => 'SearchMovieField');
jest.mock('../../infoPanel/infoPanel', () => 'InfoPanel');
jest.mock('../../../helpers/getYear', () => jest.fn());
jest.mock('../../../actions/index', () => ({
    searchFilmStart: jest.fn(),
    searchSimilarFilmStart: jest.fn()
}));

let props;

describe('FilmItem component', () => {
    beforeEach(() => {
        props = {
            dispatch: mockDigest,
            searchType: 'searchType',
            show_title: 'showTitle',
            release_year: 'release_year',
            category: 'category',
            id: 111,
            poster: 'poster'
        };
    });

    it('FilmItem should be shown correctly', () => {
        const tree = renderer.create(
            <FilmItem {...props} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('FilmItem should call \'searchFilmStart\' and \'searchSimilarFilmStart\' redux actions on link click', () => {
        const tree = shallow(<FilmItem {...props} />);
        const link = tree.find('.film-item__title-link');

        link.simulate('click');
        expect(mockDigest.mock.calls.length).toBe(2);
        expect(searchFilmStart).toHaveBeenCalledWith(111, 'searchType');
        expect(searchSimilarFilmStart).toHaveBeenCalledWith(111);
    });
});
