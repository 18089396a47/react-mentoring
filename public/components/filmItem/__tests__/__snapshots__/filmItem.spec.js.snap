// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmItem component FilmItem should be shown correctly 1`] = `
<li
  className="film-item"
>
  <Link
    onClick={[Function]}
    to="/film/111"
  >
    <img
      alt="showTitle"
      className="film-item__poster"
      src="poster"
    />
  </Link>
  <div
    className="film-item__info-wrapper"
  >
    <div
      className="film-item__title-wrapper"
    >
      <Link
        className="film-item__title-link"
        onClick={[Function]}
        to="/film/111"
      >
        <h3
          className="film-item__title"
        >
          showTitle
        </h3>
      </Link>
      <span
        className="film-item__year"
      />
    </div>
    <span
      className="film-item__category"
    >
      category
    </span>
  </div>
</li>
`;
